2019/05/5.rb

```diff
@@ -0,0 +1,169 @@
+  output_index = Integer(intcode[index+3])
+end
+
+def run_intcode(input)
+  
+  i = 0 
+  loop do
+  
+    if i == input.length
+      break
+    end
+    
+    command = input[i]
+    opcode = get_opcode_for_command(command.to_s)
+       
+    if opcode == :halt
+      
+      puts "End Program"
+      break
+      
+    elsif opcode == :add
+        
+      first_input_value = get_first_parameter_value(command, input, i)
+      second_input_value = get_second_parameter_value(command, input, i)   
+      output_index = get_third_parameter_value(command, input, i)
+      
+      result = first_input_value + second_input_value
+      input[output_index] = result.to_s
+      
+    elsif opcode == :multiply
+
+      first_input_value = get_first_parameter_value(command, input, i)
+      second_input_value = get_second_parameter_value(command, input, i)
+      output_index = get_third_parameter_value(command, input, i)
+
+      result = first_input_value * second_input_value
+      input[output_index] = result.to_s
+
+    elsif opcode == :read
+
+      first_input_value = Integer(input[i+1])
+      input[first_input_value] = 5 # first input value, this is a dirty hack
+      
+    elsif opcode == :write
+    
+      first_input_index = Integer(input[i+1])
+      puts input[first_input_index]
+
+    elsif opcode == :jump_if_true
+      first_input_value = get_first_parameter_value(command, input, i)
+      second_input_value = get_second_parameter_value(command, input, i)
+
+      if first_input_value != 0
+        i = Integer(second_input_value)
+        next
+      end
+
+    elsif opcode == :jump_if_false
+      first_input_value = get_first_parameter_value(command, input, i)
+      second_input_value = get_second_parameter_value(command, input, i)
+
+      if first_input_value == 0
+        i = Integer(second_input_value)
+        next
+      end
+
+    elsif opcode == :less_than
+      first_input_value = get_first_parameter_value(command, input, i)
+      second_input_value = get_second_parameter_value(command, input, i)
+      output_index = Integer(input[i+3])
+      
+      if first_input_value < second_input_value
+        input[output_index] = "1"
+      else
+        input[output_index] = "0"
+      end
+
+    elsif opcode == :equals
+      first_input_value = get_first_parameter_value(command, input, i)
+      second_input_value = get_second_parameter_value(command, input, i)
+      output_index = Integer(input[i+3])
+      
+      parameters_equal = first_input_value == second_input_value
+      if parameters_equal
+      	input[output_index] = "1"
+      else
+        input[output_index] = "0"
+      end
+    else
+      puts "Weird opcode (#{opcode} at #{i}), command: #{command}, aborting..."
+    end
+    
+    number_of_parameters = number_of_parameters_for_opcode(opcode)
+    
+    i += number_of_parameters
+    i += 1 # next command
+  end
+end
+
+def get_opcode_for_command(command)
```

Comment:
That `opcode` concept looks like it has a lot of potential for becoming proper class, with subclasses and such and this method would become the *factory method*. But I guess, time's too short for "proper OO" 😉

Basically everything is controlled by these symbols, so I could imagine classes like

AddCommand, MultiplyCommand, JumpIfFalse, etc.

Reply:
Yeah, you're probably right. I've fiddled around with classes in Ruby as you can see [here](https://github.com/zeitschlag/adventofcode/pull/3) or [here](https://github.com/zeitschlag/adventofcode/blob/master/2019/03/third_day.rb). I think, in the long run, it will totally make sense to rewrite this with classes.

Thank you for your suggestion ☺️